redis/cluster.py

```diff
@@ -0,0 +1,1249 @@
+    @see: http://redis.io/commands/cluster-slaves # list of string
+    """
+    resp = str_if_bytes(resp)
+    current_host = options.get('current_host', '')
+
+    def parse_slots(s):
+        slots, migrations = [], []
+        for r in s.split(' '):
+            if '->-' in r:
+                slot_id, dst_node_id = r[1:-1].split('->-', 1)
+                migrations.append({
+                    'slot': int(slot_id),
+                    'node_id': dst_node_id,
+                    'state': 'migrating'
+                })
+            elif '-<-' in r:
+                slot_id, src_node_id = r[1:-1].split('-<-', 1)
+                migrations.append({
+                    'slot': int(slot_id),
+                    'node_id': src_node_id,
+                    'state': 'importing'
+                })
+            elif '-' in r:
+                start, end = r.split('-')
+                slots.extend(range(int(start), int(end) + 1))
+            else:
+                slots.append(int(r))
+
+        return slots, migrations
+
+    resp = resp.splitlines()
+    nodes = []
+    for line in resp:
+        parts = line.split(' ', 8)
+        self_id, addr, flags, master_id, ping_sent, \
+            pong_recv, config_epoch, link_state = parts[:8]
+
+        host, ports = addr.rsplit(':', 1)
+        port, _, cluster_port = ports.partition('@')
+
+        node = {
+            'id': self_id,
+            'host': host or current_host,
+            'port': int(port),
+            'cluster-bus-port': int(
+                cluster_port) if cluster_port else 10000 + int(port),
+            'flags': tuple(flags.split(',')),
+            'master': master_id if master_id != '-' else None,
+            'ping-sent': int(ping_sent),
+            'pong-recv': int(pong_recv),
+            'link-state': link_state,
+            'slots': [],
+            'migrations': [],
+        }
+
+        if len(parts) >= 9:
+            slots, migrations = parse_slots(parts[8])
+            node['slots'], node['migrations'] = tuple(slots), migrations
+
+        nodes.append(node)
+
+    return nodes
+
+
+PRIMARY = "primary"
+REPLICA = "replica"
+ALL_PRIMARIES = "all-primaries"
+ALL_REPLICAS = "all-replicas"
+ALL_NODES = "all-nodes"
+RANDOM = "random"
+SLOT_ID = "slot-id"
+
+REDIS_ALLOWED_KEYS = (
+    "charset",
+    "connection_class",
+    "connection_pool",
+    "db",
+    "decode_responses",
+    "encoding",
+    "encoding_errors",
+    "errors",
+    "host",
+    "max_connections",
+    "redis_connect_func",
+    "password",
+    "port",
+    "retry_on_timeout",
+    "socket_connect_timeout",
+    "socket_keepalive",
+    "socket_keepalive_options",
+    "socket_timeout",
+    "ssl",
+    "ssl_ca_certs",
+    "ssl_certfile",
+    "ssl_cert_reqs",
+    "ssl_keyfile",
+    "unix_socket_path",
+    "username",
+)
+KWARGS_DISABLED_KEYS = (
+    "host",
+    "port",
+)
+
+# Not complete, but covers the major ones
+# https://redis.io/commands
+READ_COMMANDS = frozenset([
```

Comment:
I'll happily update this list for you, if you can somehow give me write access.

Reply:
I sent you a collaborator access invitation